apps/src/dance/Dance.js

```diff
@@ -487,14 +486,19 @@ Dance.prototype.preview = async function () {
   await this.nativeAPI.ensureSpritesAreLoaded(charactersReferenced);
   this.hooks.find(v => v.name === 'runUserSetup').func();
 
-  // Force preview draw to occur **after** any
-  // draw iterations already queued up.
-  // redraw() (rather than draw()) is p5's recommended way
-  // of drawing once.
-  setTimeout(() => {
+  const previewDraw = () => {
+    this.nativeAPI.setEffectsInPreviewMode(true);
```

Comment:
Oh one thing I just noticed - I think when i was looking into live preview, when I moved this line here (after running the `runUserSetup` func) some foreground effects weren't appearing because the effect `init()` calls were being made with the preview flag set to false. Is that happening for you?

Reply:
Good catch, I'll take a closer look!

Reply:
@sanchitmalhotra126 just took a look at where the init call comes from -- I made a change here such that init gets called in preview mode:

https://github.com/code-dot-org/code-dot-org/pull/54456/commits/65e0f72943bef3c79fe0430bfabbffd14f480d45

The interpreter/etc is a bit fuzzy to me, but I _think_ this should be ok? I verified that an effect that is dependent on preview mode in its init function (tacos) didn't work before, and works now.

Reply:
Cool, yeah I think that makes sense! My understanding is that code in the `whenSetup` block registers setup hook/hooks, and then `runUserSetup` runs those setup hooks which include functions like `setBackgroundEffect`/`setForegroundEffect` where the effect `init` calls actually happen?